ouroboros-network-protocols/src/Ouroboros/Network/Protocol/ObjectDiffusion/Type.hs

```diff
@@ -0,0 +1,359 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE EmptyCase #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE PatternSynonyms #-}
+
+-- | The type of the object diffusion protocol.
+--
+-- This is used to diffuse generic objects between nodes.
+module Ouroboros.Network.Protocol.ObjectDiffusion.Type
+  ( pattern InboundAgency,
+    InboundAgency,
+    pattern OutboundAgency,
+    OutboundAgency,
+    pattern AsInbound,
+    AsInbound,
+    pattern AsOutbound,
+    AsOutbound,
+    pattern ReflInboundAgency,
+    ReflInboundAgency,
+    pattern ReflOutboundAgency,
+    ReflOutboundAgency,
+    ObjectDiffusion (..),
+    Message (..),
+    SingObjectDiffusion (..),
+    SingBlockingStyle (..),
+    StBlockingStyle (..),
+    BlockingReplyList (..),
+    NumObjectIdsToAck (..),
+    NumObjectIdsToReq (..),
+    -- re-exports
+    SizeInBytes (..),
+  )
+where
+
+import Control.DeepSeq
+import Data.Kind (Type)
+import Data.List.NonEmpty (NonEmpty)
+import Data.Monoid (Sum (..))
+import Data.Singletons
+import Data.Word (Word16)
+import GHC.Generics
+import Network.TypedProtocol.Core
+import NoThunks.Class (NoThunks (..))
+import Ouroboros.Network.SizeInBytes (SizeInBytes (..))
+import Ouroboros.Network.Util.ShowProxy
+import Quiet (Quiet (..))
+
+pattern OutboundAgency :: Agency
+pattern OutboundAgency = ClientAgency
+type OutboundAgency = 'ClientAgency
+
+pattern InboundAgency :: Agency
+pattern InboundAgency = ServerAgency
+type InboundAgency = 'ServerAgency
+
+pattern AsOutbound :: PeerRole
+pattern AsOutbound = AsClient
+type AsOutbound = 'AsClient
+
+pattern AsInbound :: PeerRole
+pattern AsInbound = AsServer
+type AsInbound = 'AsServer
+
+pattern ReflOutboundAgency :: ReflRelativeAgency ClientAgency (r :: RelativeAgency) (r :: RelativeAgency)
+pattern ReflOutboundAgency = ReflClientAgency
+type ReflOutboundAgency = 'ReflClientAgency
+
+pattern ReflInboundAgency :: ReflRelativeAgency ServerAgency (r :: RelativeAgency) (r :: RelativeAgency)
+pattern ReflInboundAgency = ReflServerAgency
+type ReflInboundAgency = 'ReflServerAgency
+
+-- | The kind of the object diffusion protocol, and the types of the states in
+-- the protocol state machine.
+--
+-- We describe this protocol using the label \"inbound\" for the peer that is
+-- receiving objects, and \"inbound\" for the one sending them.
+type ObjectDiffusion :: Agency -> Type -> Type -> Type
+data ObjectDiffusion initAgency objectId object where
```

Comment:
```suggestion
type ObjectDiffusion :: Type -> Type -> Type
data ObjectDiffusion objectId object where
```
The `initAgency` doesn't seem to be used anywhere in the protocol definitions, can it be dropped?